core/node/l1_recovery/src/l1_fetcher/l1_fetcher.rs

```diff
@@ -0,0 +1,1083 @@
+    PriorityOpId(PriorityOpId),
+    None,
+}
+
+impl L1Fetcher {
+    pub fn new(config: L1FetcherConfig, eth_client: Box<DynClient<L1>>) -> Result<Self> {
+        tracing::info!("Initialized L1 fetcher with config: {config:?}");
+        Ok(L1Fetcher { eth_client, config })
+    }
+
+    fn hyperchain_contract() -> Contract {
+        hyperchain_contract()
+    }
+
+    fn commit_functions() -> Result<Vec<Function>> {
+        let contract = Self::hyperchain_contract();
+        Ok(vec![
+            contract.functions_by_name("commitBatches").unwrap()[0].clone(),
+            contract
+                .functions_by_name("commitBatchesSharedBridge")
+                .unwrap()[0]
+                .clone(),
+        ])
+    }
+
+    fn rollup_event_by_type(event_type: RollupEventType) -> Result<Event> {
+        let event_name = match event_type {
+            RollupEventType::Commit => "BlockCommit",
+            RollupEventType::Prove => "BlocksVerification",
+            RollupEventType::Execute => "BlockExecution",
+            RollupEventType::NewPriorityTx => "NewPriorityRequest",
+        };
+        Ok(Self::hyperchain_contract().events_by_name(event_name)?[0].clone())
+    }
+
+    fn extract_l1_batch_number_from_rollup_event_log(
+        event_type: RollupEventType,
+        log: &Log,
+    ) -> L1BatchNumber {
+        let topic_index = match event_type {
+            RollupEventType::Commit => 1,
+            RollupEventType::Prove => 2,
+            RollupEventType::Execute => 1,
+            _ => panic!("{event_type:?} event doesn't have l1_batch_number"),
+        };
+        L1BatchNumber(log.topics[topic_index].to_low_u64_be() as u32)
+    }
+
+    fn extract_priority_op_id_from_rollup_event_log(
+        event_type: RollupEventType,
+        log: &Log,
+    ) -> PriorityOpId {
+        assert_eq!(event_type, RollupEventType::NewPriorityTx);
+        L1Tx::try_from(log.clone()).unwrap().common_data.serial_id
+    }
+
+    pub async fn find_block_near_genesis(&self) -> U64 {
+        let mut start_block: U64 = U64::zero();
+        let last_l1_block_number = L1Fetcher::get_last_l1_block_number(&self.eth_client)
+            .await
+            .unwrap();
+        let mut end_block = last_l1_block_number;
+        // we use step override as such big step value guarantees that we don't miss
+        // events on non-local networks
+        let step = 50_000;
+        while start_block + 2 * step < end_block {
+            let mid = (start_block + end_block) / 2;
+            let logs = self
+                .get_logs(mid, mid + step - 1, RollupEventType::Commit)
+                .await;
+            if !logs.is_empty() {
+                end_block = mid;
+            } else {
+                start_block = mid;
+            }
+        }
+        end_block = last_l1_block_number;
+        let mut current_block = start_block;
+        loop {
+            let filter_to_block = cmp::min(current_block + step - 1, end_block);
```

Comment:
fyi, instead of `cmp::min` you can use `min` method (no imports required), for example
`let filter_to_block = end_block.min(...)`.